8.0 Migration of account_constraints and account_default_draft_move

adrienpeiffer opened the pull request:
Port account default draft move and account constraint to 8.0

Commit:
[ADD] Migrate account_constraint and account_default_draft_move to V8

Commit:
[FIX] Fix depends

Commit:
[IMP] Better modularity and naming convention

pedrobaeza commented:
Why are you including also _account_constraints_ in this PR?

Commit:
[IMP] Replace active by auto_install key and set to False

adrienpeiffer commented:
@pedrobaeza Because account_default_draft_move depends on account_constraints

eLBati commented:
About travis configuration, please see https://github.com/OCA/account-financial-tools/pull/36

pedrobaeza commented:
Ah, OK. Is there any test we can run to check functionality?

Commit:
[IMP] Improve modularity and migrate account_move_line

yvaucher commented on account_constraints/model/account_move_line.py:
```diff
@@ -0,0 +1,158 @@
+                    return True
+                err_msg = (_('Invoice name (id): %s (%s)') %
+                            (line.invoice.name, line.invoice.id))
+                raise models.except_orm(
```
Wouldn't it be better to import directly openerp.exceptions and use directly exceptions.except_orm ?

yvaucher commented on account_default_draft_move/account.py:
```diff
@@ -0,0 +1,62 @@
+        res = super(AccountInvoice, self).action_move_create()
+        for inv in self:
+            if inv.move_id:
+                inv.move_id.write({'state': 'draft'})
```
here you can write it:
inv.move_id.state = 'draft'

Commit:
[IMP] Improve use of new api for account_default_draft_move

Commit:
[IMP] Use exceptions import on account_default_draft_move

Commit:
[IMP] Use exceptions import on account_constraints

yvaucher commented on account_constraints/model/account_move_line.py:
```diff
@@ -0,0 +1,159 @@
+                                                   check=check)
+
+    @api.cr_uid_ids_context
+    def write(self, cr, uid, ids, vals, context=None, check=True,
```
You could use @api.multi here to remove cr, uid, ids and context

Then you could access context with: self._context

guewen replied:
@yvaucher I prefer `self.env.context` because `self._context` accesses to a private attribute.

adrienpeiffer replied:
@guewen  @yvaucher  unlink and write method of account_move_line isn't yet migrated in new API and unlink attemps to write in context which is frozen.This is the major reason why these methods don't use the new API.
Also context argument isn't at the end of method declaration of these methods and this causes errors.

yvaucher commented on account_constraints/model/account_move_line.py:
```diff
@@ -0,0 +1,159 @@
+        return True
+
+    @api.cr_uid_ids_context
+    def unlink(self, cr, uid, ids, context=None, check=True):
```
You could use @api.multi here to remove cr, uid, ids and context

Then you could access context with: self._context

pedrobaeza replied:
Better to use self.env.context, as this is the official way. The other variable may dissapear or change its meaning as it is a private one.

adrienpeiffer replied:
@yvaucher  unlink method of account_move_line isn't yet migrated in new API and this one attemps to write in context which is frozen.This is the major reason why these methods don't use the new API.

Commit:
[TEST] Add tests for account default draft move addons

Commit:
[TEST] Add tests for account constraints addons

eLBati commented:
Diff between 7.0 and 8.0:
[account_default_draft_move](http://pastie.org/9548247)
[account_constraints](http://pastie.org/9548256)

sbidoul commented:
Anyone willing to review this one?

sbidoul committed:
[IMP] allow unreconciling bank statement lines by adding the cancel button

Since account_constraint forbids the deletion of account moves that
are linked to a bank statement, we add the possibility to delete them
from the bank statement itself, to provide a way to fix erroneous matches.

sbidoul commented:
Since account_constraint forbids the deletion of account moves that are linked to a bank statement, I added the possibility to delete them from the bank statement view, to provide a way to fix erroneous matches.

sbidoul commented:
@fclementic2c would you like to test this one?

adrienpeiffer commented:
Thanks @sbidoul

bwrsandman commented:
Needs are rebase

sbidoul commented:
@bwrsandman rebased

guewen commented on account_default_draft_move/__init__.py:
```diff
@@ -19,3 +19,4 @@
 ##############################################################################
 from . import account
 from . import account_bank_statement
+from . import tests
```
You do not need to import tests, they will be searched for by the test runner

guewen commented on account_default_draft_move/account.py:
```diff
@@ -0,0 +1,63 @@
+
+from openerp import models, api
+from openerp import exceptions
+from openerp.tools.translate import _
```
it is also importable from openerp, so the 3 lines could be reduced to:

```
from openerp import models, api, exceptions, _
```

guewen commented on account_default_draft_move/account.py:
```diff
@@ -0,0 +1,63 @@
+                continue
+            else:
+                if not line.journal_id.update_posted:
+                    raise exceptions.except_orm(
```
```
raise exceptions.Warning(_('You cannot modify a posted entry of this journal.'
                           'First you should set the journal to allow'
                           ' cancelling entries.'))
```

guewen commented on account_default_draft_move/account.py:
```diff
@@ -0,0 +1,63 @@
+        if self:
+            self._cr.execute('UPDATE account_move '
+                             'SET state=%s '
+                             'WHERE id IN %s', ('draft', self._ids,))
```
better to use `self.ids`

guewen commented on account_default_draft_move/account_bank_statement.py:
```diff
@@ -0,0 +1,32 @@
+        res = super(AccountBankStatement, self).button_confirm_bank()
+        for st in self:
+            for st_line in st.line_ids:
+                st_line.journal_entry_id.state = 'draft'
```
I think it would be better to do the write on all the entries at once (1 UPDATE), something like

```
entries = self.mapped('line_ids.journal_entry_id')
entries.write({'state': 'draft'})
```

(not 100% sure for the syntax of `mapped`, to check)

adrienpeiffer replied:
@guewen This syntax looks correct. Thanks !

guewen commented on account_default_draft_move/tests/__init__.py:
```diff
@@ -0,0 +1,38 @@
+
+checks = [
+    test_account_default_draft_move,
+]
```
`fast_suite` and `checks` are no longer used in 8.0

guewen commented on account_constraints/__init__.py:
```diff
@@ -17,4 +17,5 @@
 ##############################################################################
-import account_constraints
+import model
+from . import tests
```
You do not need to import tests, they will be searched for by the test runner

guewen commented on account_constraints/__openerp__.py:
```diff
@@ -51,8 +51,8 @@
-  through the parent object to make his modification.
+  that the user cannot make mistakes even in draft state, he must pass through
+  the parent object to make his modification.
 
```
If the description is moved to a file named `README.rst`, it will be displayed on GitHub and on odoo

guewen commented on account_constraints/model/account_move.py:
```diff
@@ -0,0 +1,42 @@
+        (_check_fiscal_year,
+            'You cannot create entries with date not in the '
+            'fiscal year of the chosen period',
+            ['line_id']),
```
Can be replaced by `@api.constrains('line_id')` on the method (and the method has to raise `exceptions.Warning` if the check is failed). `'line_id'` seems wrong to me though, shouldn't it be `'journal_id', 'period_id', 'date'`?

adrienpeiffer replied:
@guewen I agree with you

guewen commented on account_constraints/model/account_move_line.py:
```diff
@@ -0,0 +1,159 @@
+                err_msg = (_('Invoice name (id): %s (%s)') %
+                            (line.invoice.name, line.invoice.id))
+                raise exceptions.except_orm(
+                    _('Error'),
```
Can be replaced by `exceptions.Warning`

guewen commented on account_constraints/model/account_move_line.py:
```diff
@@ -0,0 +1,159 @@
+                err_msg = (_('Bank statement name (id): %s (%s)') %
+                            (line.statement_id.name, line.statement_id.id))
+                raise exceptions.except_orm(
+                    _('Error'),
```
Can be replaced by exceptions.Warning

guewen commented on account_constraints/model/account_move_line.py:
```diff
@@ -0,0 +1,159 @@
+        (_check_currency_company,
+         "You can't provide a secondary currency if "
+         "the same than the company one.",
+         ['currency_id']),
```
Can be replaced by `@api.constrains` on the methods

guewen commented:
I added a few comments, it is not required to correct them all, but it allows me to share some things I have learned on the new API, I hope they will be useful.

tafaRU commented:
@guewen your comments are for sure useful,  thanks for sharing them!

Commit:
[IMP] not necessary to import test folder on init file

Commit:
[IMP] Group import

Commit:
[IMP] Use constraint decorator on account_constraints+

Commit:
[IMP] Replace except_orm with Warning

Commit:
[IMP] Use mapped method

Commit:
[IMP] Not necessary to use checks and fast suite for tests in 8.0

Commit:
[IMP] Use constraint decorator on account_constraints

adrienpeiffer commented:
@guewen Thanks for your comments !

guewen commented:
Thanks @adrienpeiffer !
:+1:

sbidoul commented:
:+1: as well

The only possible annoyance is that the cancel button (yellow arrow) on bank statement line will appear twice if account_cancel is installed). But since one goal of account_default_draft_move is to make account_cancel unncessary that should not be an issue.
